File: templates/py/protocols/mission_protocol.py
from __future__ import annotations

import asyncio
from collections.abc import Callable
from dataclasses import dataclass
from datetime import timedelta

from mavlink import (
    CommandAck,
    CommandInt,
    CommandLong,
    MavCmd,
    MavComponent,
    MavMissionResult,
    MavResult,
    MissionAck,
    MissionClearAll,
    MissionCount,
    MissionItem,
    MissionItemInt,
    MissionRequest,
    MissionRequestInt,
    MissionRequestList,
    MissionSetCurrent,
)
from mavlink_frame import MavlinkFrame
from mavlink_message import MavlinkMessage

from .command_protocol import CommandProtocol
from .mavlink_cancellation import MavlinkCancellationToken
from .mavlink_session import MavlinkSession


class MissionItems:
    @staticmethod
    def waypoint(
        seq: int,
        latitude: float,
        longitude: float,
        altitude: float,
        target_system: int,
        target_component: int,
        command: MavCmd = MavCmd.MAV_CMD_NAV_WAYPOINT,
        frame=None,
        mission_type=None,
        param1: float = 0,
        param2: float = 0,
        param3: float = 0,
        param4: float = 0,
        current: int = 0,
        autocontinue: int = 1,
    ) -> MissionItemInt:
        from mavlink import MavFrame, MavMissionType  # noqa: PLC0415

        resolved_frame = frame or MavFrame.MAV_FRAME_GLOBAL_RELATIVE_ALT_INT
        resolved_mission_type = mission_type or MavMissionType.MAV_MISSION_TYPE_MISSION
        return MissionItemInt(
            param1=param1,
            param2=param2,
            param3=param3,
            param4=param4,
            x=int(latitude * 1e7),
            y=int(longitude * 1e7),
            z=altitude,
            seq=seq,
            command=command,
            target_system=target_system,
            target_component=target_component,
            frame=resolved_frame,
            current=current,
            autocontinue=autocontinue,
            mission_type=resolved_mission_type,
        )

    @staticmethod
    def to_legacy_item(item: MissionItemInt) -> MissionItem:
        return MissionItem(
            param1=item.param1,
            param2=item.param2,
            param3=item.param3,
            param4=item.param4,
            x=item.x / 1e7,
            y=item.y / 1e7,
            z=item.z,
            seq=item.seq,
            command=item.command,
            target_system=item.target_system,
            target_component=item.target_component,
            frame=item.frame,
            current=item.current,
            autocontinue=item.autocontinue,
            mission_type=item.mission_type,
        )

    @staticmethod
    def from_legacy_item(item: MissionItem) -> MissionItemInt:
        return MissionItemInt(
            param1=item.param1,
            param2=item.param2,
            param3=item.param3,
            param4=item.param4,
            x=int(item.x * 1e7),
            y=int(item.y * 1e7),
            z=item.z,
            seq=item.seq,
            command=item.command,
            target_system=item.target_system,
            target_component=item.target_component,
            frame=item.frame,
            current=item.current,
            autocontinue=item.autocontinue,
            mission_type=item.mission_type,
        )

    @staticmethod
    def with_sequential_seq(items: list[MissionItemInt]) -> list[MissionItemInt]:
        return [
            MissionItemInt(
                param1=item.param1,
                param2=item.param2,
                param3=item.param3,
                param4=item.param4,
                x=item.x,
                y=item.y,
                z=item.z,
                seq=index,
                command=item.command,
                target_system=item.target_system,
                target_component=item.target_component,
                frame=item.frame,
                current=item.current,
                autocontinue=item.autocontinue,
                mission_type=item.mission_type,
            )
            for index, item in enumerate(items)
        ]


MissionUploadProgressCallback = Callable[[int, int, MissionItemInt], None]
MissionDownloadProgressCallback = Callable[[int, int, MissionItemInt], None]


@dataclass(frozen=True, slots=True)
class MissionSetCurrentResult:
    sequence: int
    command_ack: CommandAck | None = None


class MissionProtocol:
    def __init__(
        self,
        session: MavlinkSession,
        target_system: int,
        target_component: int,
        item_timeout: timedelta = timedelta(seconds=3),
        operation_timeout: timedelta = timedelta(seconds=10),
    ) -> None:
        self.session = session
        self.target_system = target_system
        self.target_component = target_component
        self.item_timeout = item_timeout
        self.operation_timeout = operation_timeout

    async def upload(
        self,
        items: list[MissionItemInt],
        mission_type=None,
        on_progress: MissionUploadProgressCallback | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ):
        from mavlink import MavMissionType  # noqa: PLC0415

        if cancel is not None:
            cancel.throw_if_cancelled()

        resolved_mission_type = mission_type or MavMissionType.MAV_MISSION_TYPE_MISSION
        plan = MissionItems.with_sequential_seq(items)

        await self.session.send(
            MissionCount(
                count=len(plan),
                target_system=self.target_system,
                target_component=self.target_component,
                mission_type=resolved_mission_type,
            )
        )

        for item in plan:
            if cancel is not None:
                cancel.throw_if_cancelled()

            request = await self.session.wait_for_message(
                predicate=lambda message: self._is_item_request(
                    message, item.seq, resolved_mission_type
                ),
                from_system_id=self.target_system,
                timeout=self.item_timeout,
                cancel=cancel,
            )

            if isinstance(request, MissionRequestInt):
                await self.session.send(item)
            elif isinstance(request, MissionRequest):
                await self.session.send(MissionItems.to_legacy_item(item))

            if on_progress is not None:
                on_progress(item.seq + 1, len(plan), item)

        ack = await self.session.wait_for_message_type(
            MissionAck,
            from_system_id=self.target_system,
            timeout=self.operation_timeout,
            cancel=cancel,
        )
        return ack.type

    async def download(
        self,
        mission_type=None,
        on_progress: MissionDownloadProgressCallback | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> list[MissionItemInt]:
        from mavlink import MavMissionType  # noqa: PLC0415

        if cancel is not None:
            cancel.throw_if_cancelled()

        resolved_mission_type = mission_type or MavMissionType.MAV_MISSION_TYPE_MISSION

        await self.session.send(
            MissionRequestList(
                target_system=self.target_system,
                target_component=self.target_component,
                mission_type=resolved_mission_type,
            )
        )

        count_message = await self.session.wait_for_message_type(
            MissionCount,
            from_system_id=self.target_system,
            timeout=self.operation_timeout,
            cancel=cancel,
        )

        items: list[MissionItemInt] = []

        for seq in range(count_message.count):
            if cancel is not None:
                cancel.throw_if_cancelled()

            await self.session.send(
                MissionRequestInt(
                    seq=seq,
                    target_system=self.target_system,
                    target_component=self.target_component,
                    mission_type=resolved_mission_type,
                )
            )

            item_message = await self.session.wait_for_message(
                predicate=lambda message: (
                    (
                        isinstance(message, MissionItemInt)
                        and message.seq == seq
                        and message.mission_type == resolved_mission_type
                    )
                    or (
                        isinstance(message, MissionItem)
                        and message.seq == seq
                        and message.mission_type == resolved_mission_type
                    )
                ),
                from_system_id=self.target_system,
                timeout=self.item_timeout,
                cancel=cancel,
            )

            if isinstance(item_message, MissionItemInt):
                item = item_message
            else:
                item = MissionItems.from_legacy_item(item_message)  # type: ignore[arg-type]

            items.append(item)
            if on_progress is not None:
                on_progress(len(items), count_message.count, item)

        await self.session.send(
            MissionAck(
                target_system=self.target_system,
                target_component=self.target_component,
                type=MavMissionResult.MAV_MISSION_ACCEPTED,
                mission_type=resolved_mission_type,
            )
        )

        return items

    async def clear(
        self,
        mission_type=None,
        cancel: MavlinkCancellationToken | None = None,
    ):
        from mavlink import MavMissionType  # noqa: PLC0415

        resolved_mission_type = mission_type or MavMissionType.MAV_MISSION_TYPE_MISSION

        await self.session.send(
            MissionClearAll(
                target_system=self.target_system,
                target_component=self.target_component,
                mission_type=resolved_mission_type,
            )
        )

        ack = await self.session.wait_for_message_type(
            MissionAck,
            from_system_id=self.target_system,
            timeout=self.operation_timeout,
            cancel=cancel,
        )
        return ack.type

    async def set_current(
        self, seq: int, cancel: MavlinkCancellationToken | None = None
    ) -> None:
        if cancel is not None:
            cancel.throw_if_cancelled()
        await self.session.send(
            MissionSetCurrent(
                seq=seq,
                target_system=self.target_system,
                target_component=self.target_component,
            )
        )

    async def set_current_with_command(
        self,
        seq: int,
        command: CommandProtocol | None = None,
        also_send_command: bool = True,
        reset_mission: bool = False,
        cancel: MavlinkCancellationToken | None = None,
    ) -> MissionSetCurrentResult:
        if cancel is not None:
            cancel.throw_if_cancelled()
        await self.set_current(seq, cancel=cancel)

        ack = None
        if also_send_command and command is not None:
            ack = await command.set_mission_current(
                seq, reset_mission=reset_mission, cancel=cancel
            )

        return MissionSetCurrentResult(sequence=seq, command_ack=ack)

    def _is_item_request(
        self, message: MavlinkMessage, seq: int, mission_type
    ) -> bool:
        if isinstance(message, MissionRequestInt):
            return message.seq == seq and message.mission_type == mission_type
        if isinstance(message, MissionRequest):
            return message.seq == seq and message.mission_type == mission_type
        return False


class MissionServer:
    def __init__(
        self,
        session: MavlinkSession,
        initial_mission: list[MissionItemInt] | None = None,
        mission_type=None,
    ) -> None:
        from mavlink import MavMissionType  # noqa: PLC0415

        self.session = session
        self.mission_type = mission_type or MavMissionType.MAV_MISSION_TYPE_MISSION
        self._items: list[MissionItemInt] = list(initial_mission or [])
        self._incoming: dict[int, MissionItemInt] = {}
        self._incoming_count: int | None = None
        self._frame_task = asyncio.create_task(self._listen_frames())

    @property
    def items(self) -> list[MissionItemInt]:
        return list(self._items)

    async def close(self) -> None:
        self._frame_task.cancel()
        try:
            await self._frame_task
        except asyncio.CancelledError:
            pass

    def replace_mission(self, items: list[MissionItemInt]) -> None:
        self._items = MissionItems.with_sequential_seq(items)
        self._incoming.clear()
        self._incoming_count = None

    async def _listen_frames(self) -> None:
        async for frame in self.session.frames():
            await self._on_frame(frame)

    async def _on_frame(self, frame: MavlinkFrame) -> None:
        message = frame.message

        if isinstance(message, MissionCount) and self._targets_us(message):
            if message.mission_type != self.mission_type:
                return
            self._incoming_count = message.count
            self._incoming.clear()
            if message.count > 0:
                await self._request_upload_item(frame, 0)
            else:
                await self._send_upload_ack(frame)
            return

        if isinstance(message, MissionItemInt) and self._targets_us(message):
            if message.mission_type != self.mission_type:
                return
            await self._store_incoming_item(frame, message)
            return

        if isinstance(message, MissionItem) and self._targets_us(message):
            if message.mission_type != self.mission_type:
                return
            await self._store_incoming_item(frame, MissionItems.from_legacy_item(message))
            return

        if isinstance(message, MissionRequestInt) and self._targets_us(message):
            await self._send_requested_item(frame, message.seq)
            return

        if isinstance(message, MissionRequest) and self._targets_us(message):
            await self._send_requested_item(frame, message.seq)
            return

        if isinstance(message, MissionRequestList) and self._targets_us(message):
            if message.mission_type != self.mission_type:
                return
            await self.session.send(
                MissionCount(
                    count=len(self._items),
                    target_system=frame.system_id,
                    target_component=frame.component_id,
                    mission_type=self.mission_type,
                )
            )
            return

        if isinstance(message, MissionClearAll) and self._targets_us(message):
            if message.mission_type != self.mission_type:
                return
            self._items.clear()
            self._incoming.clear()
            self._incoming_count = None
            await self.session.send(
                MissionAck(
                    target_system=frame.system_id,
                    target_component=frame.component_id,
                    type=MavMissionResult.MAV_MISSION_ACCEPTED,
                    mission_type=self.mission_type,
                )
            )

    async def _store_incoming_item(self, frame: MavlinkFrame, item: MissionItemInt) -> None:
        self._incoming[item.seq] = item
        expected = self._incoming_count
        if expected is None:
            return

        if len(self._incoming) < expected:
            await self._request_upload_item(frame, item.seq + 1)
            return

        self._items = [self._incoming[index] for index in range(expected)]
        self._incoming.clear()
        self._incoming_count = None
        await self._send_upload_ack(frame)

    async def _request_upload_item(self, request_frame: MavlinkFrame, seq: int) -> None:
        await self.session.send(
            MissionRequestInt(
                seq=seq,
                target_system=request_frame.system_id,
                target_component=request_frame.component_id,
                mission_type=self.mission_type,
            )
        )

    async def _send_upload_ack(self, request_frame: MavlinkFrame) -> None:
        await self.session.send(
            MissionAck(
                target_system=request_frame.system_id,
                target_component=request_frame.component_id,
                type=MavMissionResult.MAV_MISSION_ACCEPTED,
                mission_type=self.mission_type,
            )
        )

    async def _send_requested_item(self, request_frame: MavlinkFrame, seq: int) -> None:
        if seq < 0 or seq >= len(self._items):
            await self.session.send(
                MissionAck(
                    target_system=request_frame.system_id,
                    target_component=request_frame.component_id,
                    type=MavMissionResult.MAV_MISSION_INVALID_SEQUENCE,
                    mission_type=self.mission_type,
                )
            )
            return

        await self.session.send(self._items[seq])

    def _targets_us(self, message: MavlinkMessage) -> bool:
        target_system = getattr(message, "target_system", None)
        target_component = getattr(message, "target_component", None)
        if target_system is None or target_component is None:
            return False
        return self._matches_target(target_system, target_component)

    def _matches_target(self, target_system: int, target_component: int) -> bool:
        if target_system != self.session.system_id and target_system != 0:
            return False
        if (
            target_component != self.session.component_id
            and target_component != MavComponent.MAV_COMP_ID_ALL
        ):
            return False
        return True

File: templates/ts/protocols/mission_protocol.ts
import {
  CommandAck,
  MavCmd,
  MavComponent,
  MavFrame,
  MavMissionResult,
  MavMissionType,
  MavResult,
  MissionAck,
  MissionClearAll,
  MissionCount,
  MissionItem,
  MissionItemInt,
  MissionRequest,
  MissionRequestInt,
  MissionRequestList,
  MissionSetCurrent,
  type CommandInt,
  type CommandLong,
} from '../mavlink';
import type { MavlinkFrame } from '../mavlink_frame';
import { MavlinkMessage } from '../mavlink_message';
import type { CommandProtocol } from './command_protocol';
import { MavlinkCancellationToken } from './mavlink_cancellation';
import { MavlinkSession } from './mavlink_session';

/** Helpers for building and converting mission plan items. */
export class MissionItems {
  private constructor() {}

  /** Build a global waypoint using scaled integer lat/lon (MAVLink convention). */
  static waypoint(options: {
    seq: number;
    latitude: number;
    longitude: number;
    altitude: number;
    targetSystem: number;
    targetComponent: number;
    command?: MavCmd;
    frame?: MavFrame;
    missionType?: MavMissionType;
    param1?: number;
    param2?: number;
    param3?: number;
    param4?: number;
    current?: number;
    autocontinue?: number;
  }): MissionItemInt {
    return new MissionItemInt(
      options.param1 ?? 0,
      options.param2 ?? 0,
      options.param3 ?? 0,
      options.param4 ?? 0,
      Math.round(options.latitude * 1e7),
      Math.round(options.longitude * 1e7),
      options.altitude,
      options.seq,
      options.command ?? MavCmd.MAV_CMD_NAV_WAYPOINT,
      options.targetSystem,
      options.targetComponent,
      options.frame ?? MavFrame.MAV_FRAME_GLOBAL_RELATIVE_ALT_INT,
      options.current ?? 0,
      options.autocontinue ?? 1,
      options.missionType ?? MavMissionType.MAV_MISSION_TYPE_MISSION,
    );
  }

  /** Convert a [MissionItemInt] to the legacy [MissionItem] representation. */
  static toLegacyItem(item: MissionItemInt): MissionItem {
    return new MissionItem(
      item.param1,
      item.param2,
      item.param3,
      item.param4,
      item.x / 1e7,
      item.y / 1e7,
      item.z,
      item.seq,
      item.command,
      item.targetSystem,
      item.targetComponent,
      item.frame,
      item.current,
      item.autocontinue,
      item.missionType,
    );
  }

  /** Convert a legacy [MissionItem] to [MissionItemInt]. */
  static fromLegacyItem(item: MissionItem): MissionItemInt {
    return new MissionItemInt(
      item.param1,
      item.param2,
      item.param3,
      item.param4,
      Math.round(item.x * 1e7),
      Math.round(item.y * 1e7),
      item.z,
      item.seq,
      item.command,
      item.targetSystem,
      item.targetComponent,
      item.frame,
      item.current,
      item.autocontinue,
      item.missionType,
    );
  }

  /** Re-number items sequentially starting from zero. */
  static withSequentialSeq(items: MissionItemInt[]): MissionItemInt[] {
    return items.map((item, index) =>
      new MissionItemInt(
        item.param1,
        item.param2,
        item.param3,
        item.param4,
        item.x,
        item.y,
        item.z,
        index,
        item.command,
        item.targetSystem,
        item.targetComponent,
        item.frame,
        item.current,
        item.autocontinue,
        item.missionType,
      ),
    );
  }
}

export type MissionUploadProgressCallback = (
  sent: number,
  total: number,
  item: MissionItemInt,
) => void;

export type MissionDownloadProgressCallback = (
  received: number,
  total: number,
  item: MissionItemInt,
) => void;

/** Result of [MissionProtocol.setCurrentWithCommand]. */
export interface MissionSetCurrentResult {
  readonly sequence: number;
  readonly commandAck?: CommandAck;
}

/** GCS-side MAVLink mission protocol client. */
export class MissionProtocol {
  readonly session: MavlinkSession;
  readonly targetSystem: number;
  readonly targetComponent: number;
  readonly itemTimeoutMs: number;
  readonly operationTimeoutMs: number;

  constructor(options: {
    session: MavlinkSession;
    targetSystem: number;
    targetComponent: number;
    itemTimeoutMs?: number;
    operationTimeoutMs?: number;
  }) {
    this.session = options.session;
    this.targetSystem = options.targetSystem;
    this.targetComponent = options.targetComponent;
    this.itemTimeoutMs = options.itemTimeoutMs ?? 3000;
    this.operationTimeoutMs = options.operationTimeoutMs ?? 10_000;
  }

  /** Upload a mission plan to the vehicle. */
  async upload(
    items: MissionItemInt[],
    options: {
      missionType?: MavMissionType;
      onProgress?: MissionUploadProgressCallback;
      cancel?: MavlinkCancellationToken;
    } = {},
  ): Promise<MavMissionResult> {
    const missionType =
      options.missionType ?? MavMissionType.MAV_MISSION_TYPE_MISSION;
    options.cancel?.throwIfCancelled();
    const plan = MissionItems.withSequentialSeq(items);

    await this.session.send(
      new MissionCount(
        plan.length,
        this.targetSystem,
        this.targetComponent,
        missionType,
      ),
    );

    for (const item of plan) {
      options.cancel?.throwIfCancelled();

      const request = await this.session.waitForMessage({
        predicate: (message) =>
          MissionProtocol.isItemRequest(message, item.seq, missionType),
        fromSystemId: this.targetSystem,
        timeoutMs: this.itemTimeoutMs,
        cancel: options.cancel,
      });

      if (MavlinkMessage.isMessageOf<MissionRequestInt>(request, MissionRequestInt)) {
        await this.session.send(item);
      } else if (MavlinkMessage.isMessageOf<MissionRequest>(request, MissionRequest)) {
        await this.session.send(MissionItems.toLegacyItem(item));
      }

      options.onProgress?.call(undefined, item.seq + 1, plan.length, item);
    }

    const ack = await this.waitForMissionAck(options.cancel);

    return ack.type;
  }

  /** Download a mission plan from the vehicle. */
  async download(
    options: {
      missionType?: MavMissionType;
      onProgress?: MissionDownloadProgressCallback;
      cancel?: MavlinkCancellationToken;
    } = {},
  ): Promise<MissionItemInt[]> {
    const missionType =
      options.missionType ?? MavMissionType.MAV_MISSION_TYPE_MISSION;
    options.cancel?.throwIfCancelled();

    await this.session.send(
      new MissionRequestList(
        this.targetSystem,
        this.targetComponent,
        missionType,
      ),
    );

    const countMessage = await this.session.waitForMessageType(MissionCount, {
      fromSystemId: this.targetSystem,
      timeoutMs: this.operationTimeoutMs,
      cancel: options.cancel,
    });

    const items: MissionItemInt[] = [];

    for (let seq = 0; seq < countMessage.count; seq++) {
      options.cancel?.throwIfCancelled();

      await this.session.send(
        new MissionRequestInt(
          seq,
          this.targetSystem,
          this.targetComponent,
          missionType,
        ),
      );

      const itemMessage = await this.session.waitForMessage({
        predicate: (message) => {
          if (MavlinkMessage.isMessageOf<MissionItemInt>(message, MissionItemInt)) {
            return (
              message.seq === seq && message.missionType === missionType
            );
          }
          if (MavlinkMessage.isMessageOf<MissionItem>(message, MissionItem)) {
            return (
              message.seq === seq && message.missionType === missionType
            );
          }
          return false;
        },
        fromSystemId: this.targetSystem,
        timeoutMs: this.itemTimeoutMs,
        cancel: options.cancel,
      });

      const item = MavlinkMessage.isMessageOf<MissionItemInt>(
        itemMessage,
        MissionItemInt,
      )
        ? itemMessage
        : MissionItems.fromLegacyItem(itemMessage as MissionItem);

      items.push(item);
      options.onProgress?.call(undefined, items.length, countMessage.count, item);
    }

    await this.session.send(
      new MissionAck(
        this.targetSystem,
        this.targetComponent,
        MavMissionResult.MAV_MISSION_ACCEPTED,
        missionType,
      ),
    );

    return items;
  }

  /** Clear all mission items of the given type on the vehicle. */
  async clear(
    options: {
      missionType?: MavMissionType;
      cancel?: MavlinkCancellationToken;
    } = {},
  ): Promise<MavMissionResult> {
    const missionType =
      options.missionType ?? MavMissionType.MAV_MISSION_TYPE_MISSION;
    await this.session.send(
      new MissionClearAll(
        this.targetSystem,
        this.targetComponent,
        missionType,
      ),
    );

    const ack = await this.waitForMissionAck(options.cancel);

    return ack.type;
  }

  /** Set the active mission item sequence number on the vehicle. */
  async setCurrent(
    seq: number,
    cancel?: MavlinkCancellationToken,
  ): Promise<void> {
    cancel?.throwIfCancelled();
    await this.session.send(
      new MissionSetCurrent(seq, this.targetSystem, this.targetComponent),
    );
  }

  /** Send [MissionSetCurrent] and optionally [CommandProtocol.setMissionCurrent]. */
  async setCurrentWithCommand(
    seq: number,
    options: {
      command?: CommandProtocol;
      alsoSendCommand?: boolean;
      resetMission?: boolean;
      cancel?: MavlinkCancellationToken;
    } = {},
  ): Promise<MissionSetCurrentResult> {
    options.cancel?.throwIfCancelled();
    await this.setCurrent(seq, options.cancel);

    let commandAck: CommandAck | undefined;
    if (
      (options.alsoSendCommand ?? true) &&
      options.command !== undefined
    ) {
      commandAck = await options.command.setMissionCurrent(seq, {
        resetMission: options.resetMission,
        cancel: options.cancel,
      });
    }

    return { sequence: seq, commandAck };
  }

  private async waitForMissionAck(
    cancel?: MavlinkCancellationToken,
  ): Promise<MissionAck> {
    return this.session.waitForMessage({
      predicate: (message) => {
        if (!MavlinkMessage.isMessageOf<MissionAck>(message, MissionAck)) {
          return false;
        }
        return (
          message.targetSystem === this.session.systemId &&
          (message.targetComponent === this.session.componentId ||
            message.targetComponent === MavComponent.MAV_COMP_ID_ALL)
        );
      },
      fromSystemId: this.targetSystem,
      timeoutMs: this.operationTimeoutMs,
      cancel,
    }) as Promise<MissionAck>;
  }

  private static isItemRequest(
    message: MavlinkMessage,
    seq: number,
    missionType: MavMissionType,
  ): boolean {
    if (MavlinkMessage.isMessageOf<MissionRequestInt>(message, MissionRequestInt)) {
      return message.seq === seq && message.missionType === missionType;
    }
    if (MavlinkMessage.isMessageOf<MissionRequest>(message, MissionRequest)) {
      return message.seq === seq && message.missionType === missionType;
    }
    return false;
  }
}

/** Vehicle-side mission protocol handler for embedding in autopilot code. */
export class MissionServer {
  readonly session: MavlinkSession;
  readonly missionType: MavMissionType;

  private readonly _items: MissionItemInt[] = [];
  private readonly _incoming = new Map<number, MissionItemInt>();
  private _incomingCount: number | null = null;
  private readonly _unsubscribe: () => void;

  constructor(options: {
    session: MavlinkSession;
    initialMission?: MissionItemInt[];
    missionType?: MavMissionType;
  }) {
    this.session = options.session;
    this.missionType =
      options.missionType ?? MavMissionType.MAV_MISSION_TYPE_MISSION;
    if (options.initialMission !== undefined) {
      this._items.push(...options.initialMission);
    }
    const subscription = this.session.listenMessage((message, frame) => {
      void this._onFrame(frame, message);
    });
    this._unsubscribe = () => subscription.cancel();
  }

  get items(): readonly MissionItemInt[] {
    return this._items;
  }

  async close(): Promise<void> {
    this._unsubscribe();
  }

  replaceMission(items: MissionItemInt[]): void {
    this._items.length = 0;
    this._items.push(...MissionItems.withSequentialSeq(items));
    this._incoming.clear();
    this._incomingCount = null;
  }

  private async _onFrame(
    frame: MavlinkFrame,
    message: MavlinkMessage,
  ): Promise<void> {
    if (MavlinkMessage.isMessageOf<MissionCount>(message, MissionCount) && this._targetsUs(message)) {
      if (message.missionType !== this.missionType) {
        return;
      }
      this._incomingCount = message.count;
      this._incoming.clear();
      if (message.count > 0) {
        await this._requestUploadItem(frame, 0);
      } else {
        await this._sendUploadAck(frame);
      }
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionItemInt>(message, MissionItemInt) && this._targetsUs(message)) {
      if (message.missionType !== this.missionType) {
        return;
      }
      await this._storeIncomingItem(frame, message);
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionItem>(message, MissionItem) && this._targetsUs(message)) {
      if (message.missionType !== this.missionType) {
        return;
      }
      await this._storeIncomingItem(frame, MissionItems.fromLegacyItem(message));
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionRequestInt>(message, MissionRequestInt) && this._targetsUs(message)) {
      await this._sendRequestedItem(frame, message.seq);
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionRequest>(message, MissionRequest) && this._targetsUs(message)) {
      await this._sendRequestedItem(frame, message.seq);
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionRequestList>(message, MissionRequestList) && this._targetsUs(message)) {
      if (message.missionType !== this.missionType) {
        return;
      }
      await this.session.send(
        new MissionCount(
          this._items.length,
          frame.systemId,
          frame.componentId,
          this.missionType,
        ),
      );
      return;
    }

    if (MavlinkMessage.isMessageOf<MissionClearAll>(message, MissionClearAll) && this._targetsUs(message)) {
      if (message.missionType !== this.missionType) {
        return;
      }
      this._items.length = 0;
      this._incoming.clear();
      this._incomingCount = null;
      await this.session.send(
        new MissionAck(
          frame.systemId,
          frame.componentId,
          MavMissionResult.MAV_MISSION_ACCEPTED,
          this.missionType,
        ),
      );
    }
  }

  private async _storeIncomingItem(
    frame: MavlinkFrame,
    item: MissionItemInt,
  ): Promise<void> {
    this._incoming.set(item.seq, item);
    const expected = this._incomingCount;
    if (expected === null) {
      return;
    }

    if (this._incoming.size < expected) {
      await this._requestUploadItem(frame, item.seq + 1);
      return;
    }

    this._items.length = 0;
    for (let index = 0; index < expected; index++) {
      const stored = this._incoming.get(index);
      if (stored !== undefined) {
        this._items.push(stored);
      }
    }
    this._incoming.clear();
    this._incomingCount = null;
    await this._sendUploadAck(frame);
  }

  private async _requestUploadItem(
    requestFrame: MavlinkFrame,
    seq: number,
  ): Promise<void> {
    await this.session.send(
      new MissionRequestInt(
        seq,
        requestFrame.systemId,
        requestFrame.componentId,
        this.missionType,
      ),
    );
  }

  private async _sendUploadAck(requestFrame: MavlinkFrame): Promise<void> {
    await this.session.send(
      new MissionAck(
        requestFrame.systemId,
        requestFrame.componentId,
        MavMissionResult.MAV_MISSION_ACCEPTED,
        this.missionType,
      ),
    );
  }

  private async _sendRequestedItem(
    requestFrame: MavlinkFrame,
    seq: number,
  ): Promise<void> {
    if (seq < 0 || seq >= this._items.length) {
      await this.session.send(
        new MissionAck(
          requestFrame.systemId,
          requestFrame.componentId,
          MavMissionResult.MAV_MISSION_INVALID_SEQUENCE,
          this.missionType,
        ),
      );
      return;
    }

    await this.session.send(this._items[seq]!);
  }

  private _targetsUs(message: MavlinkMessage): boolean {
    const target = MissionServer.readTarget(message);
    if (target === null) {
      return false;
    }
    return this._matchesTarget(target.system, target.component);
  }

  private static readTarget(
    message: MavlinkMessage,
  ): { system: number; component: number } | null {
    if (
      MavlinkMessage.isMessageOf(message, MissionCount) ||
      MavlinkMessage.isMessageOf(message, MissionItemInt) ||
      MavlinkMessage.isMessageOf(message, MissionItem) ||
      MavlinkMessage.isMessageOf(message, MissionRequestInt) ||
      MavlinkMessage.isMessageOf(message, MissionRequest) ||
      MavlinkMessage.isMessageOf(message, MissionRequestList) ||
      MavlinkMessage.isMessageOf(message, MissionClearAll)
    ) {
      const targeted = message as MissionCount;
      return {
        system: targeted.targetSystem,
        component: targeted.targetComponent,
      };
    }
    return null;
  }

  private _matchesTarget(targetSystem: number, targetComponent: number): boolean {
    if (targetSystem !== this.session.systemId && targetSystem !== 0) {
      return false;
    }
    if (
      targetComponent !== this.session.componentId &&
      targetComponent !== MavComponent.MAV_COMP_ID_ALL
    ) {
      return false;
    }
    return true;
  }
}
